contributors/design-proposals/api-machinery/customresources-validation.md

```diff
@@ -381,6 +381,29 @@ Note:
 
 2. For migration of CRDs with no validation to CRDs with validation, we can create a controller that will validate and annotate invalid CRs once the spec changes, so that the custom controller can choose to delete them (this is also essentially the status condition of the CRD). This can be achieved, but it is not part of the proposal.
 
+### Admission Webhook
+
+Custom resource definitions use the the same REST endpoint as built-in Kubernetes API objects. Consequentaly standard
+[dynamic web-hook admission controllers](https://kubernetes.io/docs/admin/extensible-admission-controllers/#external-admission-webhooks) 
+can be used for validation of custom resource definitions in addition to built-in resources.
+
+However, because the creation of admission controllers is a fairly high-privilege activity,
+and in many cases the creator of a `CustomResourceDefinition` is third-party extension code
+(e.g. an instance of the _operator pattern_) which should not have the ability to
+create arbitrary admission controllers, it is necessary to have an alternate solution for
+adding web-hook validation to custom resource definitions.
```

Comment:
This raises the question about how to create+distribute the necessary certificates. Currently, we don't have a mechanism to auto-create and inject 

- the webhook server key+certificate (trusted by the apiextensions-apiserver)
- and the kube-apiserver CA (trusted and verified by the webhook).

We need both. I don't want to see every CRD controller using this implementing its own certificate generation code, and even worse creating something insecure. 

In OpenShift we have a controller that's listening on a custom annotation on apiserver pods and then creates a secret on-demand. This gives a smooth and secure setup with zero config.

Reply:
We need to solve the certificate problem. I had a brief chat with @deads2k a few weeks ago about extracting the service signing certificate controller from OpenShift and making it generally available. Whether we do that or find another solution, I believe we need to do something. Otherwise, I feel it's too much of a burden to ask users to figure out certificate generation, signing, etc. to be able to do things like this.

Reply:
I think @ncdc is saying this, but I think this is a general problem with dynamic webhooks, not specific to this proposal. We need to make it much easier for people to build dynamic webhooks in general.

Reply:
Added some text about this.